File: apps/ios/Sources/Litter/Models/PathDisplay.swift
import Foundation

/// Convert filesystem paths to short, user-facing strings.
///
/// For **local** codex paths, rewrites the app-container Documents home
/// and the app's `NSTemporaryDirectory()` to `~` and `/tmp` so the UI
/// shows `~/projects/foo` and `/tmp/x.txt` instead of
/// `/var/mobile/Containers/Data/Application/<UUID>/Documents/home/codex/...`.
///
/// For **remote** paths, delegates to the existing `abbreviateHomePath`
/// which shortens `/Users/<user>/<subpath>` and `/home/<user>/<subpath>`
/// to `~/<subpath>`.
enum PathDisplay {
    /// Callers pass `isLocal = true` only when `raw` is a path on the
    /// in-process iOS codex. Remote-server paths (SSH/WebSocket) go
    /// through `abbreviateHomePath`.
    static func display(_ raw: String, isLocal: Bool) -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return isLocal ? "~" : trimmed }
        guard isLocal else { return abbreviateHomePath(trimmed) }
        let home = HomeAnchor.path
        if trimmed == home { return "~" }
        if trimmed.hasPrefix(home + "/") {
            return "~/" + trimmed.dropFirst(home.count + 1)
        }
        let tmp = realTmp()
        if !tmp.isEmpty {
            if trimmed == tmp { return "/tmp" }
            if trimmed.hasPrefix(tmp + "/") {
                return "/tmp/" + trimmed.dropFirst(tmp.count + 1)
            }
        }
        return trimmed
    }

    /// Inverse of `display` for the local case. Accepts user-entered
    /// display strings (`~/foo`, `/tmp/x`) and produces an absolute path
    /// on the iOS sandbox. No-op for remote paths.
    static func expand(_ display: String, isLocal: Bool) -> String {
        guard isLocal else { return display }
        if display == "~" { return HomeAnchor.path }
        if display.hasPrefix("~/") {
            return HomeAnchor.path + "/" + display.dropFirst(2)
        }
        let tmp = realTmp()
        if !tmp.isEmpty {
            if display == "/tmp" { return tmp }
            if display.hasPrefix("/tmp/") {
                return tmp + "/" + display.dropFirst(5)
            }
        }
        return display
    }

    private static func realTmp() -> String {
        let raw = NSTemporaryDirectory()
        // Strip trailing slash so comparisons are uniform with prefix
        // matching that adds `/`.
        if raw.hasSuffix("/") { return String(raw.dropLast()) }
        return raw
    }
}

File: apps/android/app/src/main/java/com/litter/android/state/PathDisplay.kt
package com.litter.android.state

import android.content.Context

/**
 * Convert filesystem paths to short, user-facing strings.
 *
 * For **local** codex paths, rewrites the `HomeAnchor.path(context)` and
 * `$TMPDIR` to `~` and `/tmp` so the UI shows `~/projects/foo` and
 * `/tmp/x.txt` instead of the raw `/data/user/0/com.sigkitten.litter/files/...`
 * absolute paths.
 *
 * For **remote** paths, leaves the path as-is (remote filesystems have
 * their own natural home dirs like `/home/<user>`).
 */
object PathDisplay {
    /**
     * Callers pass [isLocal] `= true` only when [raw] is a path on the
     * in-process Android codex. Remote-server paths pass through unchanged.
     */
    fun display(raw: String, isLocal: Boolean, context: Context): String {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return if (isLocal) "~" else trimmed
        if (!isLocal) return trimmed
        val home = HomeAnchor.path(context)
        if (trimmed == home) return "~"
        if (trimmed.startsWith("$home/")) return "~/" + trimmed.substring(home.length + 1)
        val tmp = realTmp()
        if (tmp.isNotEmpty()) {
            if (trimmed == tmp) return "/tmp"
            if (trimmed.startsWith("$tmp/")) return "/tmp/" + trimmed.substring(tmp.length + 1)
        }
        return trimmed
    }

    /** Inverse of [display] for user-entered display strings on the local server. */
    fun expand(display: String, isLocal: Boolean, context: Context): String {
        if (!isLocal) return display
        if (display == "~") return HomeAnchor.path(context)
        if (display.startsWith("~/")) return HomeAnchor.path(context) + "/" + display.substring(2)
        val tmp = realTmp()
        if (tmp.isNotEmpty()) {
            if (display == "/tmp") return tmp
            if (display.startsWith("/tmp/")) return "$tmp/" + display.substring(5)
        }
        return display
    }

    private fun realTmp(): String {
        // Set by `Java_com_litter_android_core_bridge_UniffiInit_nativeBridgeInit`
        // at JNI boot. Strip trailing slash so comparisons are uniform.
        val raw = System.getenv("TMPDIR") ?: return ""
        return if (raw.endsWith("/")) raw.dropLast(1) else raw
    }
}
